### ISSUE TITLE
Dat-cli crashes syncing

### ISSUE THREAD

**Issue by icarito** [datproject/dat#856]
Here's my output on a Ubuntu 16.04 machine running nodejs v4.2.6:
```
icarito@sunjammer:~/public_html/Workspace$ dat sync .                                                                                         │···············
(node) Buffer.set is deprecated. Use array indexes instead.                                                                                   │···············
dat v13.8.1                                                                                                                                   │···············
Downloading dat: 136 files (2.4 MB)                                                                                                           │···············
                                                                                                                                              │···············
1 connection | Download 0 B/s Upload 0 B/s                                                                                                    │···············
                   7 files (2.4 MB)                                                                                                           │···············
dat synced, waiting8 files (2.4 MB)                                                                                                           │···············
[=======================0 B/s Upload 0 B/s                                                                                                    │···············
0 connections | Download 595 B/s Upload 0 B/s                                                                                                 │···············
dat synced, waiting for updates.                                                                                                              │···············
Downloading updates...                                                                                                                        │···············
                                               1.22%                                                                                          │···············
                                                                                                                                              │···············
                                                                                                                                              │···············
                                                                                                                                              │···············
/usr/local/lib/node_modules/dat/src/lib/download.js:33                                                                                        │···············
      debug('archive sync', state.stats.get())                                                                                                │···············
      ^                                                                                                                                       │···············
                                                                                                                                              │···············
RangeError: Maximum call stack size exceeded                                                                                                  │···············
    at Hyperdrive.<anonymous> (/usr/local/lib/node_modules/dat/src/lib/download.js:33:7)                                                      │···············
    at emitNone (events.js:67:13)                                                                                                             │···············
    at Hyperdrive.emit (events.js:166:7)                                                                                                      │···············
    at /usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:141:14                                                                │···············
    at kick (/usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:236:5)                                                          │···············
    at Object.callback (/usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:223:7)                                               │···············
    at Peer._downloadRange (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:368:13)                                   │···············
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:299:12)                                           │···············
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:300:51)                                           │···············
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:300:51)                                           │···············
```

**Comment by icarito** [comment]
and fwiw my dat is at https://github.com/somosazucar/Jappy/tree/master/workspace

**Comment by joehand** [comment]
Wasn't able to reproduce on node v4.2.6 with syncing to a fresh directory.

This may be related to live updates, was the source dat updating while you were running sync? Or was there any other non-trivial things you were doing?

**Comment by icarito** [comment]
It's likely the `dat` updated during sync, but not intensively or repeatedly. I thought sync was meant to run indefinitively? I just updated to v6.11 and so far so good.

**Comment by joehand** [comment]
>  I just updated to v6.11 and so far so good.

Cool! 

> It's likely the dat updated during sync, but not intensively or repeatedly. I thought sync was meant to run indefinitively?

Yes, sync should run indefinitely. The error stack just seemed related to a live update so I wanted to see if that was the case and why I couldn't reproduce.

**Comment by icarito** [comment]
Thanks, to describe the likely filesystem update case, my app writes to the filesystem two or three, four files each time (it's an editor, every open file), and then one more file one second later (compilation output).

**Comment by icarito** [comment]
Okay so I've had a crash now with node 6.11.2:
```
icarito@sunjammer:~/public_html/Workspace$ dat sync .
dat v13.8.1
Downloading dat: 153 files (2.5 MB)

1 connection | Download 135 B/s Upload 0 B/s

dat synced, waiting for updates.
[==========================================] 102.79%


Ctrl+C to Exit
/usr/local/lib/node_modules/dat/src/lib/download.js:32
    archive.on('sync', function () {
                                ^

RangeError: Maximum call stack size exceeded
    at Hyperdrive.<anonymous> (/usr/local/lib/node_modules/dat/src/lib/download.js:32:33)
    at emitNone (events.js:86:13)
    at Hyperdrive.emit (events.js:185:7)
    at /usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:141:14
    at kick (/usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:236:5)
    at Object.callback (/usr/local/lib/node_modules/dat/node_modules/hyperdrive/index.js:223:7)
    at Peer._downloadRange (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:368:13)
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:299:12)
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:300:51)
    at Peer.update (/usr/local/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:300:51)
```

**Comment by millette** [comment]
I've seen the same **Max call stack size** error 2-3 times over a month or two. Last time was with dat://019d05c89e9469d401730172b6eb52e7880c1550fda1720671f76c7229a80bae (140 MiB in 15,000 files). Using node 8.5.0.

**Comment by millette** [comment]
A slightly different error this time:

```
/home/millette/n/lib/node_modules/dat/node_modules/sparse-bitfield/index.js:38
Bitfield.prototype.getByte = function (i) {
                                      ^

RangeError: Maximum call stack size exceeded
    at Bitfield.getByte (/home/millette/n/lib/node_modules/dat/node_modules/sparse-bitfield/index.js:38:39)
    at Iterator.seek (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/bitfield.js:209:36)
    at Iterator.random (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/bitfield.js:215:16)
    at Peer._downloadRange (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:373:43)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:311:12)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:312:51)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:312:51)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:312:51)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:312:51)
    at Peer.update (/home/millette/n/lib/node_modules/dat/node_modules/hypercore/lib/replicate.js:312:51)
```

Running the same dat version 13.8.1 everywhere but various node versions.
The one instance that failed twice today is also off by one on the number of files, showing one less than reality, when another peer shows the correct number of files (and hasn't failed yet today).

**Comment by creationix** [comment]
I've seen this a lot as well, often when trying to mirror a dat that's updating another machine.

**Comment by mafintosh** [comment]
Try reinstalling, it might now be fixed by https://github.com/mafintosh/hypercore/commit/01c7a7e73aa3906d572e4516f11267c74979e494

**Comment by millette** [comment]
Reinstalled on 3 machines yesterday, no failures since.

**Comment by millette** [comment]
Pretty sure this is fixed with the 13.8.2 release.

**Comment by icarito** [comment]
Thanks for this fix! I confirm my dat's been running fine for a couple of months now :-)
